### PR TITLE
Render Image Data

### PR DESCRIPTION
**Pull request by stefwalter**
This starts to render image data onto its own page. The URLs contain the tags of an ImageStream such as:

```
/project/my-project/browse/images/busybox:1.23
```

The actual rendering code comes from a shared component, shared with the Atomic Registry.

While this currently works, it's not ready to be merged. Remaining work:

 * [x] Create new repository for bower component (@danmcp) "openshift/registry-image-widgets"
 * [x] Register bower component, remove WIP bower_components commit. (@stefwalter)
 * [x] Use tabs to switch between sections on the browser page.
 * [x] Fix general styling
 * [x] Use non-cockpit code to convert bytes to 'MB'
 * [x] Decide how to render 'To pull this image' with appropriate registry host. (Just placeholder for now).
 * [x] Fixups for neview

CC @jwforres

### CODE REVIEW

**Comment by stefwalter** [comment]
Shared bower component is temporarily here: https://github.com/stefwalter/image-widgets/

**Comment by stefwalter** [comment]
@jwforres Are you okay with the bower and github name ```kubernetes-image-widgets``` or something else would be better? I'm thinking forward a bit and hoping that we'll have first class image representations in Kube upstream at some point.

@jwforres How do we get the actual public registry host name that users can use to pull the image onto their machine. When it's not available we substitute a "registry" placeholder.

**Comment by stefwalter** [comment]
[Video](https://www.youtube.com/watch?v=oN5RsX0tFpc)

**Comment by stefwalter** [comment]
Cockpit integration of these same widgets is blocking on https://github.com/cockpit-project/cockpit/pull/4771 ... another pull request to follow on that end.

**Comment by jwforres** [comment]
@stefwalter for the naming of the bower component, i hesitate to call it kubernetes-image-widgets, i dont think there are any plans to upstream imagestreams but @smarterclayton can weigh in here

**Comment by jwforres** [comment]
what was your reason for leaving the : as the delimeter in the URL?  I think i'd rather stick with URL conventions and add the path delimeter so there is a distinction between the image stream page and the image stream tag page, then you wouldnt have to do the special logic in the routes

File: app/scripts/app.js [hunk 1]
```diff
@@ -123,7 +124,13 @@ angular
         controller: 'ImagesController'
       })
       .when('/project/:project/browse/images/:image', {
-        templateUrl: 'views/browse/image.html',
+        templateUrl: function(params) {
+          if (params.image && params.image.indexOf(':') === -1) {
```
Comment (jwforres) [inline app/scripts/app.js]: See https://github.com/openshift/origin-web-console/pull/235#issuecomment-235908594
would rather not have to do this

**Comment by jwforres** [comment]
I noticed the *.map files are being included in the dist of the image widget bower component, I'd prefer those not be included in what we are shipping, it will bloat the final built bindata.go

**Comment by jwforres** [comment]
Ah sorry, realized you temporarily have the bower component embedded here.  As long as the *.map files aren't in the "main" of the bower.json spec of the component then it won't be an issue.

**Comment by jwforres** [comment]
On the config tab where you have the command, I understand that its the command the container will be run with, but I think there needs to be something to tell our users that.

**Comment by jwforres** [comment]
So when we show the container command elsewhere we use the `<code>` style, like what you have for pulling the image, is there a reason for using a different style for the container's command than for the docker pull command?

debugterminal

**Comment by jwforres** [comment]
@smarterclayton @deads2k see @stefwalter 's question about the registry https://github.com/openshift/origin-web-console/pull/235#issuecomment-235542107

how can we know in the console where to point them to as an end user, is it possible today or are we stuck using a placeholder?

**Comment by deads2k** [comment]
> how can we know in the console where to point them to as an end user, is it possible today or are we stuck using a placeholder?

The question is how can i find the hostname is assigned to a route that points to our internal docker registry?  There is no way for a normal user to retrieve that information via the API today.

**Comment by stefwalter** [comment]
@jwforres I'll call the bower component "registry-image-widgets" ... but can you make a new repository called "image-widgets" here: https://github.com/kubernetes-ui

Or is the hosting at kubernetes-ui also a problem? I think certain of the widgets can be used with raw docker manifests, history, layers ... so although this is currently very openshift focused, it can be used without openshift for several cases.

**Comment by jwforres** [comment]
Lets host it under the openshift org, we have a few other bower components
hosted there now too  @danmcp can create the repo   is there a reason to
not call the repo registry-image-widgets since thats what you are calling
the component?   so openshift/registry-image-widgets for the repo

**Comment by danmcp** [comment]
@stefwalter @jwforres https://github.com/openshift/registry-image-widgets

**Comment by stefwalter** [comment]
Thanks @danmcp

File: app/scripts/app.js (hunk 1 above)
Comment (stefwalter) [inline app/scripts/app.js]: Fixed.

**Comment by stefwalter** [comment]
> So when we show the container command elsewhere we use the <code> style, like what you have for pulling the image, is there a reason for using a different style for the container's command than for the docker pull command?

Fixed.

**Comment by stefwalter** [comment]
Ready for review.

**Comment by jwforres** [comment]
let me know when you are ready for a code review

**Comment by stefwalter** [comment]
@jwforres Should be ready now. I've run 'grunt build' on each commit.

**Comment by jwforres** [comment]
Do you have screenshots of what things look like at mobile resolution?

**Comment by stefwalter** [comment]
> This means any time there is a change to the image stream object it's going to re-fetch the ImageStreamTag. Whether that tag has been updated or not. Can we limit it to only fetching the tag again if we need to? 

Only the one that's currently displayed.  Openshift doesn't allow us to watch ImageStreamTag objects. You can reproduce this with this sorta command, which produces a 404:

```
curl http://127.0.0.1:8001/oapi/v1/watch/namespaces/marmalade/imagestreamtags
```

> I think we know per tag when it was last updated don't we?

How would we know that? We know when the tag data itself changed, but not when the referenced data changes. At least on my test systemst, I see no ```resourceVersion``` or ```uid``` in the ```status.tags``` data.

**Comment by stefwalter** [comment]
> Do you have screenshots of what things look like at mobile resolution?

screen shot 2016-07-30 at 07 19 42
screen shot 2016-07-30 at 07 19 47
screen shot 2016-07-30 at 07 19 49

**Comment by stefwalter** [comment]
> How would we know that? We know when the tag data itself changed, but not when the referenced data changes. At least on my test systems, I see no resourceVersion or uid in the status.tags data.

In addition even the ```ImageStreamTag``` object doesn't have a uid. Nor do the embedded image objects contain their ```resourceVersion``` or ```uid```.

**Comment by stefwalter** [comment]
FWIW, this change only displays rough approximations of image layers until https://github.com/openshift/origin/issues/7880 is fixed or https://github.com/openshift/origin/pull/7956 is merged. But that's not a blocker for this change.

**Comment by jwforres** [comment]
I thought there was something on the ImageStream object that said when the
last sync had happened @smarterclayton

I don't know if there is enough detail there to know that the particular
tag you are interested in has been updated and should be fetched again.

If we don't have enough info there to know when to refetch, then I would at
least want this updated to include a _.debounce so that if we get a bunch
of events across the imagestream websocket within a short time (200ms or
so?) that they don't all trigger separate http requests for the
ImageStreamTag.

**Comment by stefwalter** [comment]
> I thought there was something on the ImageStream object that said when the
last sync had happened @smarterclayton

ImageStreams are updated by lots more than a sync, no?

> If we don't have enough info there to know when to refetch, then I would at
least want this updated to include a _.debounce so that if we get a bunch
of events across the imagestream websocket within a short time (200ms or
so?) that they don't all trigger separate http requests for the
ImageStreamTag.

Added a debounce.

**Comment by jwforres** [comment]
@stefwalter on the mobile screenshots it would be good if the digest and identifier could be forced to wrap, in the console code we have a class "word-break" that does this (sets the best word-break classes for each browser)  since the images are in a separate repo though i not that class doesnt exist outside the console context.  I dont mind fixing this in a follow-up though.

**Comment by jwforres** [comment]
[merge]

**Comment by jwforres** [comment]
@stefwalter Just talked to @smarterclayton and he said there is a 'generation' on status, if that changes then that's when you should re-fetch ImageStreamTag.  Could you do that in a follow-up PR?

**Comment by stefwalter** [comment]
> @stefwalter Just talked to @smarterclayton and he said there is a 'generation' on status, if that changes then that's when you should re-fetch ImageStreamTag. Could you do that in a follow-up PR?

Sure. I'm away for the next few days, but I can follow this up later.

**Pull request merged**